### PR TITLE
Buyback with Uniswap v3

### PR DESCRIPTION
**Pull request by DanielVF**
If you made a contract change, make sure to complete the checklist below before merging it in master.

Refer to our [documentation](https://github.com/OriginProtocol/security) for more details about contract security best practices.

Contract change checklist:
  - [x] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [x] Unit tests pass
  - [x] Slither tests pass with no warning
  - [x] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

**Comment by DanielVF** [comment]
Fork testing script for swap:

```python
from addresses import *
from world import *

brownie.network.web3.provider.make_request('hardhat_impersonateAccount', [BINANCE])
brownie.network.web3.provider.make_request('hardhat_impersonateAccount', [GOVERNOR])


old_buyback = Contract("0x7294CD3C3eb4097b03E1A61EB2AD280D3dD265e6")
new_buyback = Contract.from_abi("Buyback","0xaB5E7B701B605f74AaC1b749Fd50715f0DEd7Bc5",old_buyback.abi)

# Generate a million dollars of yield
usdc.transfer(vault_core, 1e6 * 1e6, {'from': BINANCE})
vault_core.rebase({'from': GOVERNOR})

# Swap OUSD for OGN
vault_core.allocate({'from': GOVERNOR})
print(ousd.balanceOf(new_buyback) / 1e18)
print(ogn.balanceOf(new_buyback) / 1e18)
```

**Comment by DanielVF** [comment]
Chainlink mockup

```python
chain_ogn_eth = Contract.from_explorer("0x2c881b6f3f6b5ff6c975813f87a4dad0b241c15b")
chain_eth_usd = Contract.from_explorer("0x5f4ec3df9cbd43714fe2740f5e3616155c5b8419")
# Expected OGN per OUSD
expected_ogn_per_ousd = 1e44 / (chain_eth_usd.latestAnswer() * chain_ogn_eth.latestAnswer())
```

**Comment by DanielVF** [comment]
With the new Chainlink price protection, the previous sample code with a million dollars of yield in one shot now correctly skips the buyback since the slippage is too high.

The amount sold, and thus also slippage, can be reduced by transferring out some OUSD before doing the allocate:

```
new_buyback.transferToken(OUSD, 90*1e21, {'from': GOVERNOR})
```

File: contracts/contracts/buyback/Buyback.sol [hunk 1]
```diff
@@ -52,26 +68,78 @@ contract Buyback is Governable {
...
+                usdt,
+                uint24(3000), // Pool fee, usdt -> weth9
+                weth9,
+                uint24(3000), // Pool fee, weth9 -> ogn
+                ogn
+            ),
+            recipient: address(this),
+            deadline: uint256(block.timestamp + 1000),
+            amountIn: sourceAmount,
+            amountOutMinimum: uint256(0)
```
Comment (mikeshultz) [inline contracts/contracts/buyback/Buyback.sol:105]: Is this safe?  If one of those three pools gets extremely lopsided somehow, we could end up with a poor `amountOut`.  Wonder if we should add a guesstimate here to prevent something like that?

File: contracts/contracts/buyback/Buyback.sol [hunk 2]
```diff
@@ -52,26 +68,78 @@ contract Buyback is Governable {
...
+        });
+
+        uint256 recieved = UniswapV3Router(uniswapAddr).exactInput(params);
+
+        // Check
+        uint256 minExpected = expectedOgnPerOUSD(sourceAmount).mul(96).div(100);
```
Comment (mikeshultz) [inline contracts/contracts/buyback/Buyback.sol:111]: Suppose this is pretty related to the previous comment.  Why not send `minExpected` into `amountOutMinimum` rather than reverting here?

Also, does this math mean we're good with 4% slippage here?

File: contracts/contracts/buyback/Buyback.sol [hunk 3]
```diff
@@ -52,26 +68,78 @@ contract Buyback is Governable {
...
+        require(_feed != address(0), "Asset not available");
+        (
+            uint80 roundID,
+            int256 _iprice,
+            uint256 startedAt,
+            uint256 timeStamp,
+            uint80 answeredInRound
+        ) = AggregatorV3Interface(_feed).latestRoundData();
+        uint256 _price = uint256(_iprice);
+        require(_price > 0, "Price must be greater than zero");
```
Comment (mikeshultz) [inline contracts/contracts/buyback/Buyback.sol:139]: Correct me if I'm wrong because I mix them up sometimes, but I don't think `_price` can ever be negative, being unsigned.  Might want to move this `require()` up one line and check against `_iprice`, as casting a negative signed to unsigned could lead to some unexpected weirdness.

Also, maybe rename one of those vars to prevent confusion between `_price` and `_iprice` when reading?

File: contracts/contracts/buyback/Buyback.sol [hunk 4]
```diff
@@ -52,26 +68,78 @@ contract Buyback is Governable {
...
+        return uint256(_price);
```
Comment (mikeshultz) [inline contracts/contracts/buyback/Buyback.sol:140]: Redundant?

File: contracts/contracts/buyback/BuybackConstructor.sol [hunk 5]
```diff
@@ -1,15 +1,19 @@
...
 import { Governable } from "../governance/Governable.sol";
-
-import { IUniswapV2Router } from "../interfaces/uniswap/IUniswapV2Router02.sol";
+import "../interfaces/chainlink/AggregatorV3Interface.sol";
 import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import {UniswapV3Router} from "../interfaces/UniswapV3Router.sol";
 
 contract BuybackConstructor is Governable {
```
Comment (mikeshultz) [inline contracts/contracts/buyback/BuybackConstructor.sol:11]: Same comments on this one from the other contract.  

I realize you aren't adding this, but why do we have both `Buyback` and `BuybackConstructor`?

**Review by mikeshultz — COMMENTED** [review]
Took a first pass at the contracts.  Some questions inline.

File: contracts/contracts/buyback/Buyback.sol [hunk 6]
```diff
@@ -52,26 +68,78 @@ contract Buyback is Governable {
      * protocol (e.g. Sushiswap)
      **/
     function swap() external onlyVault {
+        // Don't revert everything, even if the buyback fails.
+        // We want the overall transaction to continue regardless.
+        (bool success, bytes memory data) = address(this).call(
+            abi.encodeWithSignature("swapAndCheck()")
+        );
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol:75]: Buyback is currently called as part of our large mints. If something goes with the buyback price during the large mint, I don't want the overall transaction with the mint to fail. By forcing an external call here, I can catch the revert after it happens, and the sale that happened during the external call is reverted into nothingness, but execution continues cleanly from here.

That's the theory anyway.

File: contracts/contracts/buyback/Buyback.sol (hunk 3 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol:139]: Good points here.

File: contracts/contracts/buyback/Buyback.sol (hunk 2 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol:111]: Yes, I guess if I trust the router, I can just move this expected number into the param and call it a day.

File: contracts/contracts/buyback/BuybackConstructor.sol (hunk 5 above)
Comment (DanielVF) [reply contracts/contracts/buyback/BuybackConstructor.sol:11]: Way more gas efficient to read a contract address from code than from a storage slot. But it makes tests so much better if you can set the addresses. Newer solidities provide an "immutable" slot attribute that writes the value to code during the constructor, giving you the best of both worlds, but we don't have it the 5.11 we are using.

File: contracts/contracts/buyback/Buyback.sol (hunk 1 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol:105]: Checked after the TX, but yes, might just change this to run here.

File: contracts/contracts/buyback/Buyback.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikeshultz) [reply contracts/contracts/buyback/Buyback.sol:75]: Ah right, there's no way to prevent the revert from bubbling up otherwise (though I thought I heard someone mention try/catch with modern Solidity).  Thanks

File: contracts/contracts/buyback/Buyback.sol (hunk 2 above)
Comment (mikeshultz) [reply contracts/contracts/buyback/Buyback.sol:111]: In this case you're trusting the router to return `received`, just the same.

File: contracts/contracts/buyback/Buyback.sol (hunk 3 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol]: Great, should be fixed, have a look.

File: contracts/contracts/buyback/Buyback.sol (hunk 4 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol]: Fixed

File: contracts/contracts/buyback/Buyback.sol (hunk 2 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol]: Changed.

File: contracts/contracts/buyback/Buyback.sol (hunk 1 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol]: Done

File: contracts/contracts/buyback/Buyback.sol (hunk 3 above)
Comment (mikeshultz) [reply contracts/contracts/buyback/Buyback.sol]: Nice and clean :+1:

**Review by mikeshultz — APPROVED** [review]
Thanks for addressing my feedback.

LGTM  :+1:

**Comment by mikeshultz** [comment]
Also, we should make sure the `Buyback` contract is added to analytics with relevant triggers so we know when `BuybackFailed` happens.  Probably not an event we want to miss.

**Comment by DanielVF** [comment]
#### Requirements

Buys back OGN with a percentage of OUSD yield. Resistant to uniswap economic attacks.

### Deployment Considerations

_NA_

#### Internal State

No internal state is modified during operation. Contract storage slots are configuration. All dynamic state is looked up from external sources.

💡 If we were using a more recent solidity version, we could immutable all these configuration addresses into the contract, and save gas on everything the contract does.

#### Attack

If this contract gets stupid amounts of OUSD, the slippage from swapping such a large amount can go over the slippage threshold, pausing swaps. Governance can fix this by temporarily withdrawing some OUSD.

If this were a long term issue, another option would be to cap the size that can be swapped at any given moment.

#### Logic

_LGTM_


#### Tests

  - [x] Each logical branch has a test
  - [ ] Each require() has a test
  - [ ] Edge conditions are tested

#### Flavor

_Made a big change to flavor. Looks better._

#### Overflow

- [x] Never use "+" or "-", always use safe math
- [x] Check that all for loops use uint256

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying
- [x] (If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing.)
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.


#### Authentication

- [x] Never use tx.origin
- [x] Check that every external/public function should actually be external
- [x] Check that every external/public function has the correct authentication

#### Cryptographic code

_No crypto_

#### Gas problems

_No for loops._

#### External calls

- [x] Contract addresses passed in are validated. _No user supplied addresses._
- [x] Unsafe external calls
- [x] Reentrancy guards on all state changing functions
    - [x] Still doesn't protect against external contracts changing the state of the world if they are called.
- [x] Malicious behaviors. _Only talking to trusted contracts._
- [x] Could fail from stack depth problems (low level calls must require success)
- [x] No slippage attacks (we need to validate expected tokens received)
- [x] Oracles, one of:
  - [ ] No oracles
  - [x] Oracles can't be bent _(not bent on chain, anyway)_
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Don't call balanceOf for external contracts to determine what they will do, when they instead use internal accounting?


#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [x] Contract has no payable methods.

**Comment by DanielVF** [comment]
I've fork tested this again, and it's ready for a review, @tomlinton.

File: contracts/contracts/buyback/Buyback.sol [hunk 7]
```diff
@@ -51,25 +67,70 @@ contract Buyback is Governable {
      * @dev Execute a swap of OGN for OUSD via Uniswap or Uniswap compatible
      * protocol (e.g. Sushiswap)
      **/
-    function swap() external onlyVault {
-        if (uniswapAddr == address(0)) return;
-
+    function swap() external onlyVault nonReentrant {
         uint256 sourceAmount = ousd.balanceOf(address(this));
         if (sourceAmount < 1000 * 1e18) return;
+        if (uniswapAddr == address(0)) return;
```
Comment (tomlinton) [inline contracts/contracts/buyback/Buyback.sol:73]: Not really related to this PR but might be nice to add: with the setUniswapAddr we should only run the safeApprove functions on uniswapAddr if address is not the zero address.

File: contracts/contracts/buyback/Buyback.sol [hunk 8]
```diff
@@ -51,25 +67,70 @@ contract Buyback is Governable {
...
+        // 97% should be the limits of our oracle errors.
+        // If this swap somtimes skips when it should succeed, that’s okay,
```
Comment (tomlinton) [inline contracts/contracts/buyback/Buyback.sol:75]: Typo

File: contracts/contracts/buyback/Buyback.sol [hunk 9]
```diff
@@ -51,25 +67,70 @@ contract Buyback is Governable {
...
+        // the amounts will get get sold the next time this runs,
+        // when presumably the oracles are more accurate.
```
Comment (tomlinton) [inline contracts/contracts/buyback/Buyback.sol:77]: This comment is very poetic, it feels like a haiku

**Review by tomlinton — APPROVED** [review]
👍

File: contracts/contracts/buyback/Buyback.sol (hunk 9 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol:77]: Hah, so it is!

File: contracts/contracts/buyback/Buyback.sol (hunk 7 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol:73]: Done

File: contracts/contracts/buyback/Buyback.sol (hunk 8 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol:75]: Fixed. Thanks
